crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -1540,32 +1533,44 @@ impl<InstalledPackages: InstalledPackagesProvider> ResolverState<InstalledPackag
         dependencies: impl IntoIterator<Item = &'data Requirement> + 'parameters,
         extra: Option<&'parameters ExtraName>,
         markers: &'parameters ResolverMarkers,
-        requires_python: Option<&'parameters MarkerTree>,
+        python_requirement: &'parameters PythonRequirement,
     ) -> impl Iterator<Item = Cow<'data, Requirement>> + 'parameters
     where
         'data: 'parameters,
     {
         self.overrides
             .apply(dependencies)
-            .filter(move |requirement| {
+            .filter_map(move |requirement| {
                 // If the requirement would not be selected with any Python version
                 // supported by the root, skip it.
-                if !satisfies_requires_python(requires_python, requirement) {
-                    trace!(
-                        "skipping {requirement} because of Requires-Python {requires_python:?}",
-                        // OK because this filter only applies when there is a present
-                        // Requires-Python specifier.
-                        requires_python = requires_python.unwrap()
+                let requirement = if let Some(requires_python) = python_requirement.target().and_then(|target| target.as_requires_python()).filter(|_| !requirement.marker.is_true()) {
+                    let marker = requirement.marker.clone().simplify_python_versions(
```

Comment:
Hmm.. I'm surprised this works. It's important to note that calling `simplify_python_versions` creates a *different* marker tree than the same tree parsed from the lockfile. For example, if we restrict the range of `python_version >= 3.12` given `requires-python >= 3.9`, we now get the marker tree `[3.9,3.12)=>false,[3.12,inf)=>true`. If you parse that same expression from the lockfile, you get `[0,3.12)=>false,[3.12,inf)=>true`, which is not equal. This can cause false negatives in a `--locked` operation. Current we solve this [by reapplying the requires-python restriction after parsing the lockfile](https://github.com/astral-sh/uv/blob/main/crates/uv-resolver/src/lock.rs#L82). However, I believe this PR now restricts markers to the *narrowed* python requirement, which we do not reapply, so I'm surprised we aren't seeing failing test cases. Is there something I'm missing and can we remove the extra code in `Lock::from_toml` now?

Reply:
That is... interesting. If I remove that code from `from_toml`, `lock_conditional_dependency_extra` fails (for one).

Reply:
It may just be that this _doesn't_ work (for that particular problem) and we just don't have a proper test case for it.

Reply:
Any other ideas on how we can solve this problem?

Reply:
I think we could fix this by removing the whole idea of a restricted range with a special "rest" range, and then only apply simplification when displaying a marker.

Reply:
I guess the other option is: try to write this to the lockfile.

Reply:
Can we use `environment-markers` to try and enforce this narrowing when we deserialize the lockfile?

Reply:
Maybe.. how would we map those to requirements?

Reply:
Could we instead just retain the lower-bound, like `>=3.7, <3.10` instead of using `<3.10`? Does that cause other problems?

Reply:
Ok @ibraheemdev -- I added a failing test for this.